pkg/metrics/sampler/matcher.go

```diff
@@ -0,0 +1,209 @@
+//  Copyright 2020 New Relic Corporation. All rights reserved.
+//  SPDX-License-Identifier: Apache-2.0
+
+package sampler
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/newrelic/infrastructure-agent/pkg/log"
+)
+
+var (
+	logger          = log.WithComponent("sampler-matcher")
+	typesToEvaluate = map[string]bool{"ProcessSample": true}
+)
+
+// ExpressionMatcher is an interface every evaluator must implement
+type ExpressionMatcher interface {
+	// Evaluate evaluates a property
+	Evaluate(event interface{}) bool
+}
+
+type attributeCache map[string]string
+
+var attrCache attributeCache
+
+type regexCompiledCache map[string]*regexp.Regexp
+
+var regexCache regexCompiledCache
+
+func init() {
+	attrCache = attributeCache{}
+	attrCache = make(map[string]string)
+	attrCache["process.name"] = "ProcessDisplayName"
+	attrCache["process.executable"] = "CmdLine"
+
+	regexCache = regexCompiledCache{}
+}
+
+type matcher struct {
+	PropertyName  string
+	ExpectedValue interface{}
+	Evaluator     func(expected interface{}, actual interface{}) bool
+}
+
+func (p matcher) Evaluate(event interface{}) bool {
+	skip := skipSample(event, typesToEvaluate)
+	if skip {
+		// evaluation determines if the sample is "included".
+		// returning true will make sure it is
+		return true
+	}
+
+	actualValue := getFieldValue(event, p.PropertyName)
+	if actualValue == nil {
+		return false
+	}
+	isMatch := p.Evaluator(p.ExpectedValue, actualValue)
+	logger.Debugf("'%v' matches expression '%v >> '%v': %v",
+		actualValue, p.PropertyName, p.ExpectedValue, isMatch)
+	return isMatch
+}
+
+func getFieldValue(object interface{}, fieldName string) interface{} {
+	v := reflect.ValueOf(object)
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+	fv := v.FieldByName(fieldName)
+	if fv.IsValid() && fv.CanInterface() {
+		fieldValue := fv.Interface()
+		return fieldValue
+	}
+
+	logger.Debugf("field '%v' does NOT exist in sample", fieldName)
```

Comment:
I think an entry per sample might flood the log, don't you?

there is a `trace` package that fits in here, requiring users to explicitly define which component to trace